src/envelope.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use crate::errors::{InternalPakeError, PakeError};
+use generic_array::{
+    typenum::{Unsigned, U32},
+    GenericArray,
+};
+use hkdf::Hkdf;
+use hmac::{Hmac, Mac, NewMac};
+use rand_core::{CryptoRng, RngCore};
+use sha2::{Digest, Sha256};
+
+// Constant string used as salt for HKDF computation
+const STR_ENVU: &[u8] = b"EnvU";
+
+/// The length of the "export key" output by the client registration
+/// and login finish steps
+pub(crate) type ExportKeySize = <Sha256 as Digest>::OutputSize;
+
+/// This struct is a straightforward instantiation of the trait separating the
+/// three components in Vecs
+pub(crate) struct Envelope {
+    nonce: Vec<u8>,
+    ciphertext: Vec<u8>,
+    hmac: Vec<u8>,
+}
+
+type NonceLen = U32;
+
+impl Envelope {
+    /// The additional number of bytes added to the plaintext
+    pub(crate) fn additional_size() -> usize {
+        Self::nonce_size() + <Sha256 as Digest>::OutputSize::to_usize()
+    }
+
+    fn hmac_key_size() -> usize {
+        <Sha256 as Digest>::OutputSize::to_usize()
+    }
+
+    fn hmac_size() -> usize {
```

Comment:
I suspect we can replace this (used only in `from_bytes`) with `hmac_key_size`

Reply:
Technically the hmac key and the hmac itself can be of different sizes. I will leave it as-is for now unless you disagree!

Reply:
Fair enough!